File: frontend/src/types/reports.ts
export interface AssetUtilizationItem {
  asset_id: string; asset_tag: string; asset_name: string;
  category: string | null; department: string | null;
  allocation_count: number; total_days_allocated: number; status: string;
}
export interface AssetUtilizationResponse { most_used: AssetUtilizationItem[]; idle: AssetUtilizationItem[]; }

export interface MaintenanceFrequencyItem {
  asset_id: string | null; asset_tag: string | null; asset_name: string | null;
  category: string | null; total_requests: number; completed: number; avg_cost: number | null;
}
export interface MaintenanceFrequencyResponse { by_asset: MaintenanceFrequencyItem[]; by_category: MaintenanceFrequencyItem[]; }

export interface AssetDueItem {
  asset_id: string; asset_tag: string; asset_name: string;
  category: string | null; department: string | null;
  due_type: string; due_date: string | null; days_until_due: number | null; status: string;
}

export interface DeptAllocationItem {
  department_id: string | null; department_name: string;
  total_allocations: number; active_allocations: number;
  overdue_allocations: number; unique_assets: number;
}

export interface BookingHeatmapSlot { hour: number; day_of_week: number; booking_count: number; }

export interface ReportsSummary {
  utilization: AssetUtilizationResponse;
  maintenance_frequency: MaintenanceFrequencyResponse;
  assets_due: AssetDueItem[];
  dept_allocation: DeptAllocationItem[];
  booking_heatmap: BookingHeatmapSlot[];
}

File: backend/app/schemas/reports.py
"""Reports & Analytics schemas."""
from pydantic import BaseModel
from typing import Optional


class AssetUtilizationItem(BaseModel):
    asset_id: str
    asset_tag: str
    asset_name: str
    category: Optional[str]
    department: Optional[str]
    allocation_count: int
    total_days_allocated: float
    status: str


class AssetUtilizationResponse(BaseModel):
    most_used: list[AssetUtilizationItem]
    idle: list[AssetUtilizationItem]


class MaintenanceFrequencyItem(BaseModel):
    asset_id: Optional[str]
    asset_tag: Optional[str]
    asset_name: Optional[str]
    category: Optional[str]
    total_requests: int
    completed: int
    avg_cost: Optional[float]


class MaintenanceFrequencyResponse(BaseModel):
    by_asset: list[MaintenanceFrequencyItem]
    by_category: list[MaintenanceFrequencyItem]


class AssetDueItem(BaseModel):
    asset_id: str
    asset_tag: str
    asset_name: str
    category: Optional[str]
    department: Optional[str]
    due_type: str          # "maintenance" | "warranty_expiry" | "retirement"
    due_date: Optional[str]
    days_until_due: Optional[int]
    status: str


class DeptAllocationItem(BaseModel):
    department_id: Optional[str]
    department_name: str
    total_allocations: int
    active_allocations: int
    overdue_allocations: int
    unique_assets: int


class BookingHeatmapSlot(BaseModel):
    hour: int          # 0–23
    day_of_week: int   # 0=Mon 6=Sun
    booking_count: int


class ReportsSummary(BaseModel):
    utilization: AssetUtilizationResponse
    maintenance_frequency: MaintenanceFrequencyResponse
    assets_due: list[AssetDueItem]
    dept_allocation: list[DeptAllocationItem]
    booking_heatmap: list[BookingHeatmapSlot]
